noir_stdlib/src/hash/sha256.nr

```diff
@@ -202,66 +246,209 @@ fn verify_msg_block<let N: u32>(
+    let modulo = msg_byte_ptr % INT_SIZE;
+    if modulo != 0 {
+        let msg_field = msg_block[int_byte_ptr];
+        for i in 0..INT_SIZE {
+            if i < INT_SIZE - modulo {
+                assert_eq(from_field_byte(msg_field, msg_byte_ptr), zero as u8);
+                msg_byte_ptr = msg_byte_ptr + 1;
+            }
+        }
+        int_byte_ptr = int_byte_ptr + 1;
+    }
+
+    // Check the rest of the numbers.
+    for i in 0..max_int_byte_ptr {
+        if i >= int_byte_ptr {
             assert_eq(msg_block[i], zero);
         }
     }
 }
 
+// Verify that up to the byte pointer the two blocks are equal.
+// At the byte pointer the new block can be partially zeroed.
+fn verify_msg_block_equals_last(
+    msg_block: MSG_BLOCK,
+    last_block: MSG_BLOCK,
+    mut msg_byte_ptr: BLOCK_BYTE_PTR,
+) {
+    // msg_byte_ptr is the position at which they are no longer have to be the same.
+    // First integer which is supposed to be (partially) zero contains that pointer.
+    let mut int_byte_ptr = msg_byte_ptr / INT_SIZE;
+
+    // Check partial zeroes.
+    let modulo = msg_byte_ptr % INT_SIZE;
+    if modulo != 0 {
+        // Reconstruct the partially zero field from the last block.
+        let last_field = last_block[int_byte_ptr];
+        let mut msg_field: u32 = 0;
+        // Reset to where they are still equal.
+        msg_byte_ptr = msg_byte_ptr - modulo;
+        for i in 0..INT_SIZE {
+            msg_field = msg_field * 256;
+            if i < modulo {
+                msg_field = msg_field + from_field_byte(last_field, msg_byte_ptr) as u32;
+                msg_byte_ptr = msg_byte_ptr + 1;
+            }
+        }
+        assert_eq(msg_block[int_byte_ptr], msg_field);
+    }
+
+    for i in 0..INT_SIZE_PTR {
+        if i < int_byte_ptr {
+            assert_eq(msg_block[i], last_block[i]);
+        }
+    }
+}
+
+// Pack the value of byte into a field of an existing message block, shifting the byte according to its position.
+fn set_msg_byte(mut msg_block: MSG_BLOCK, msg_byte_ptr: BLOCK_BYTE_PTR, msg_byte: u8) -> MSG_BLOCK {
+    let i = msg_byte_ptr / INT_SIZE;
+    msg_block[i] = set_field_byte(msg_block[i], msg_byte_ptr, msg_byte);
+    msg_block
+}
+
+// Replace one byte in a field.
+fn set_field_byte(msg_field: INT_FIELD, msg_byte_ptr: BLOCK_BYTE_PTR, msg_byte: u8) -> INT_FIELD {
+    let old_byte = from_field_byte(msg_field, msg_byte_ptr);
+    let old_field = into_field_byte(old_byte, msg_byte_ptr);
+    let new_field = into_field_byte(msg_byte, msg_byte_ptr);
+    msg_field - old_field + new_field
+}
+
+// Get a byte of a field in a `MSG_BLOCK` according to its overall position in the `BLOCK_SIZE` space.
+fn from_field_byte(mut msg_field: INT_FIELD, msg_byte_ptr: BLOCK_BYTE_PTR) -> u8 {
+    // How many times do we have to shift to the right to get to the position we want?
+    let max_shifts = INT_SIZE - 1;
+    let shifts = max_shifts - msg_byte_ptr % INT_SIZE;
+    for i in 0..max_shifts {
```

Comment:
Can you use an `if is_unconstrained() {}` to just use a bitshift if we're running in an unconstrained context?

Reply:
I suppose we can. Would that work on `Field` as well? Maxim said the compiler turns bit shifts into divisions, and we use bit shift in `1 << 7` in constrained code too. Is there a rule of thumb to follow?

Reply:
Done in https://github.com/noir-lang/noir/pull/6324/commits/2fa178f19eda14ddc09622bf12923a99c0ea7ece

Reply:
We can't do any bitshifts on `Field` types, only on integers.

The general rule of thumb on bitshifts is that in a constrained context we don't have a native way to perform bitshifts within a circuit. We then need to convert it into multiplications/divisions. If we know by how much we're shifting by at compile time this can be easily turned into a single multiplication/division and so is relatively cheap. If we don't know how many bits we're shifting by until runtime then we need to do a lot more work to handle potential cases and so it's more expensive, generally we want to avoid doing bitshifts of this type.

In unconstrained code however we have native support for bitshifts in the VM so there's no need to decompose the bitshifts into multiplications and so we can just do it directly. This helps in this case as we don't need to have the for-loop, etc. and the function then simplifies down to `(msg_field >> shift) as u8`

Reply:
Ah, sorry I think you misinterpreted me. The `rshift8` functions aren't what I was talking about here as those are cases where we're always shifting by a constant amount so it's basically equivalent in constrained and unconstrained code. I think we can revert the creation of `rshift8` as the multiplication we had previously was fine.

What I was referring to was this whole for-loop where we're emulating a bitshift by a number of bits known at runtime. In this case we need to break down the constrained code to perform a number of divisions as we're doing currently but the unconstrained version can do a standard bitshift.

Reply:
Gotcha. I added a `shifts` parameter to `lshift8` and `rshift8` in https://github.com/noir-lang/noir/pull/6324/commits/081c8ba6f80a8a6cb3a063dee9abcef81b7a6819 to get rid of the for loop.